src/plugins/data/common/index_patterns/types.ts

```diff
@@ -11,21 +11,77 @@ import { ToastInputFields, ErrorToastOptions } from 'src/core/public/notificatio
 // eslint-disable-next-line
 import type { SavedObject } from 'src/core/server';
 import { IFieldType } from './fields';
-import { RUNTIME_FIELD_TYPES } from './constants';
 import { SerializedFieldFormat } from '../../../expressions/common';
 import { KBN_FIELD_TYPES, DataViewField } from '..';
 import { FieldFormat } from '../../../field_formats/common';
 
 export type FieldFormatMap = Record<string, SerializedFieldFormat>;
 
-export type RuntimeType = typeof RUNTIME_FIELD_TYPES[number];
-export interface RuntimeField {
+export type RuntimeType = estypes.MappingRuntimeFieldType | 'composite';
+
+/**
+ * The RuntimeField that will be sent in the ES Query "runtime_mappings" object
+ * We extends the object until @elastic/elasticsearch supports "composite" type
+ * and its "fields" object.
+ *
+ * To simplify the consuming code we enforce the script to be `InlineScript` type (and not also `string`)
+ */
+export interface ESRuntimeField extends Omit<estypes.MappingRuntimeField, 'type' | 'script'> {
   type: RuntimeType;
-  script?: {
+  script?: estypes.InlineScript;
+  fields?: Record<
+    string,
+    {
+      // It is not recursive, we can't create a composite inside a composite.
+      type: Omit<RuntimeType, 'composite'>;
+    }
+  >;
+}
+
+/**
+ * The RuntimeField which is saved in the Data View saved object. We extend it to
+ * keep a reference to a possible parent composite object.
+ */
+export interface RuntimeField extends ESRuntimeField {
+  parentComposite?: string;
+}
+
+/**
+ * Runtime fields are like other fields when it comes to formatting or giving
+ * them a custom label. When adding a new runtime field in the Data view we allow the
+ * consumer to pass along a "format", "customLabel" or "popularity".
+ */
```

Comment:
Should "runtime field" be "enhanced runtime field"? Looks like `RuntimeField` type doesn't actually have format, customLabel and popularity.

Reply:
Now that I read it again it is not very clear. I will update the comment. 👍 

This is basically an interface for the `RuntimeField` with additional props.